AWS Lambda EFS post (#3606)
pulumi/docs: content/blog/aws-lambda-efs/index.md

1. You want to share persistent storage across various functions and would prefer to use file system storage instead of blob storage or a database.
1. You want to share persistent storage between Lambda and other compute (like containers or ECS), and would prefer to use file storage over blob storage or a database.

Storing data to the file system is particularly natural for the kind of small functions common in Lambda - no need to use the AWS SDK - just call `fs.writeFileSync()` (or equivalent for your language of choice).

## EFS + Lambda Concepts

To use an EFS file system from Lambda, we need a few things:

1. A VPC
1. An EFS File System in the VPC
1. EFS mount points in each of the subnets of our VPC
1. An EFS Access Point specifying how Lambda should mount the EFS volume
1. An AWS Lambda Function running in the same VPC and private subnets and referencing the EFS Access Point

![Lambda with EFS Architecture](lambdaefs.png)

Let's walk through how to build the required infrastructure.  We'll use Pulumi to provision the necessary resources.  You can check out the [full source code](https://github.com/pulumi/examples/tree/master/aws-ts-lambda-efs) in the Pulumi Examples.

> > Note: These examples depend on the newly added support for EFS with Lambda in the `2.10.0` release of the `pulumi-aws` package, which will be released later today.

Reviewer: I think we can nuke this note, esp. since "today" is (hopefully) now meaningless!

Reviewer 2: Yeah - this is available now - so we can remove the note now.

```ts
const vpc = new awsx.ec2.Vpc("vpc", { subnets: [{ type: "private" }, { type: "public" }] });
const subnetIds = await vpc.publicSubnetIds;
```

An EFS FileSystem is a regional service, but accessing it requires adding mount points to a VPC.  We can add those to the private subnets where our Functions will run.  Each mount target has a security group, and we can use the default security group to allow access only from other compute resources in the same security group in our VPC.  We could also use an existing FileSystem and mount targets if they have already been provisioned for another application.

```ts
const filesystem = new aws.efs.FileSystem("filesystem");
    const targets = [];
    for (let i = 0; i < subnetIds.length; i++) {
        targets.push(new aws.efs.MountTarget(`fs-mount-${i}`, {
            fileSystemId: filesystem.id,
            subnetId: subnetIds[i],
            securityGroups: [vpc.vpc.defaultSecurityGroupId],
        }));
    }
```

> To use EFS from our Lambda Function, we need to tell Lambda how to mount the NFS volume.  Specifically, what Unix user and group to use for access, and what folder and permissions to use as the root folder of the mount.  This is done using an EFS AccessPoint.

Reviewer: ```suggestion
To use EFS from our Lambda Function, we need to tell Lambda how to mount the NFS volume.  Specifically, we must specify the Unix user and group to use to enable access, and the location and permission scheme to use as the root folder of the mount.  This is done using an EFS AccessPoint.
```

Reviewer: (previous draft is incomplete sentence)

```ts
    const ap = new aws.efs.AccessPoint("ap", {
        fileSystemId: filesystem.id,
        posixUser: { uid: 1000, gid: 1000 },
        rootDirectory: { path: "/www", creationInfo: { ownerGid: 1000, ownerUid: 1000, permissions: "755" } },
    }, { dependsOn: targets });
```

> Finally, each Lambda Function we want to run will need to be in our VPC, use our default security group, and mount the EFS file system via our AccessPoint.  We can write a small function that creates Lambda Functions with this configuration.  We can specify where we want to mount this into the Lambda Function - though we must mount it underneath `/mnt/`.

Reviewer: ```suggestion
Finally, each Lambda Function we want to run will need to be in our VPC, use our default security group, and mount the EFS file system via our AccessPoint.  We can write a small function that creates Lambda Functions with this configuration and we can specify where we want to mount this into the Lambda Function, though we must mount it underneath `/mnt/`.
```

        return new aws.lambda.CallbackFunction(name, {
            policies: [aws.iam.ManagedPolicies.AWSLambdaVPCAccessExecutionRole, aws.iam.ManagedPolicies.AWSLambdaFullAccess],
            vpcConfig: {
                subnetIds: vpc.privateSubnetIds,
                securityGroupIds: [vpc.vpc.defaultSecurityGroupId],
            },
            fileSystemConfigs: [{ arn: ap.arn, localMountPath: "/mnt/storage" }],
            callback: f,
        });
    }
```

We now have our infrastructure all set up to start using Lambda with EFS.  Let’s build an app!

## Interact with EFS from an HTTP API

For a first simple Lambda application, we'll expose an HTTP API that GETs files from, and POSTs files to, our EFS file system.  We'll implement this with API Gateway and two routes.  Using the helper we defined above, each of these is just a few lines of code.

Note that we can use our normal Node.js `fs.readFileSync` and `fs.writeFileSync` APIs to interact with the EFS file system - no need for an AWS SDK to talk to EFS. This makes it easy to take existing software and libraries that can interact with the filesystem and run them inside Lambda.

> > Note: Yes, exposing this to the internet is insecure - it is recommended that you [add authentication or use a private endpoint](https://www.pulumi.com/docs/guides/crosswalk/aws/api-gateway/#controlling-and-managing-access-to-apis) before running this in production!

Reviewer: ```suggestion
> Note: Yes, exposing this to the internet is insecure - it is strongly recommended that you [add authentication or use a private endpoint](https://www.pulumi.com/docs/guides/crosswalk/aws/api-gateway/#controlling-and-managing-access-to-apis) before running this in production!
```

Reviewer: I really, really don't think we should post insecure examples -- ever. It's way, way too common for folks to cut/paste things they see on our blogs.
